
# Assemble mtDNA

## KMC

```bash
cd /home/hpcshared/Databases/mtdna/1.clean
k=91
k=101
k=111
k=121
k=31
k=17
declare -a specarr=("A126" "A38" "A3" "A40" "A45" "A4" "A5" "A68" "A6" "A73" "A7" "A8" "B116" "B117" "B118" "B119" "B120" "B68" "B99")
for idv in ${specarr[@]}; do mkdir tmp.$idv.k$k; kmc -k$k -fq -cs2500 @$idv.fq.lst $idv.k$k tmp.$idv.k$k;rmdir tmp.$idv.k$k; kmc_tools transform $idv.k$k histogram $idv.k$k.histo -cx2500; done
```

```bash
histfile="AA/B116.kmers.histo"
histfile="AA/B117.kmers.histo"
histfile="AA/B118.kmers.histo"
histfile="AA/B119.kmers.histo"
histfile="AA/B120.kmers.histo"
sed -i 's/\s\+/\t/g;' $histfile
perl -lane 'print if (scalar(@F)!=2);' $histfile
perl -lane 'BEGIN{$mind=500;$maxd=1000;$maxc=0;$out=0;} if ($F[0]<=$maxd and $F[0]>=$mind) {if ($F[1]>$maxc){$out=$F[0];$maxc=$F[1];}}END {print "Summit: $out";}' $histfile
```

```bash
#AA summit 
## B116	IX-0944	595
./kmer.plot.Rscript AA/B116.kmers.histo 3 1500 IX-0944 595
./kmer.plot.ribbon.Rscript AA/B116.kmers.histo 3 1500 IX-0944 282 595 921

## B117	IX-0946	469
./kmer.plot.Rscript AA/B117.kmers.histo 3 1500 IX-0946 769
./kmer.plot.ribbon.Rscript AA/B117.kmers.histo 3 1500 IX-0946 431 769 1068
## B118	IX-0957 657
./kmer.plot.Rscript AA/B118.kmers.histo 3 1000 IX-0957 326
./kmer.plot.ribbon.Rscript AA/B118.kmers.histo 3 1000 IX-0957 210 326 488
## B119	IX-0969 504
./kmer.plot.Rscript AA/B119.kmers.histo 3 1000 IX-0969 418
./kmer.plot.ribbon.Rscript k17/AA/B119.kmers.histo 3 1000 IX-0969 254 418 615
./kmer.plot.ribbon3.Rscript k17/AA/B119.kmers.histo 3 1000 "AA IX-0969" 254 418 615 20
## B120	IX-0976 704
./kmer.plot.Rscript AA/B120.kmers.histo 3 1500 IX-0976 704
./kmer.plot.ribbon.Rscript AA/B120.kmers.histo 3 1500 IX-0969 442 704 1109
```


##assembly 

### AABB

```
mtDNA=/home/hpcusers/master/huilin_hu/Mt_assemble/Real/mtDNA
mkdir ${OUTDIR}/AABB/A8
time spades.py   -k 47,63,89,99,107   -1 ${mtDNA}/AABB/A8_FDSW21H003586-1r_1.clean.fq.gz  -2  ${mtDNA}/AABB/A8_FDSW21H003586-1r_2.clean.fq.gz -o ${OUTDIR}/AABB/A8
mkdir ${OUTDIR}/AABB/A7
time spades.py   -k 47,63,89,99,107   -1 ${mtDNA}/AABB/A7_FDSW210333328-1r_1.clean.fq.gz  -2  ${mtDNA}/AABB/A7_FDSW210333328-1r_2.clean.fq.gz -o ${OUTDIR}/AABB/A7
mkdir ${OUTDIR}/AABB/A4
time spades.py   -k 47,63,89,99,107   -1 ${mtDNA}/AABB/A4_FDSW21H002742-1r_1.clean.fq.gz  -2  ${mtDNA}/AABB/A4_FDSW21H002742-1r_2.clean.fq.gz -o ${OUTDIR}/AABB/A4
mkdir ${OUTDIR}/AABB/A40
time spades.py   -k 47,63,89,99,107   -1 ${mtDNA}/AABB/A40_FDSW210333330-1r_1.clean.fq.gz  -2  ${mtDNA}/AABB/A40_FDSW210333330-1r_2.clean.fq.gz -o ${OUTDIR}/AABB/A40
```

### assembly AABBDD / AA / DD

> cmds refer to AABB but replace the fastq files



## Error correction

```bash
checkNeighbour () {
	CNpos1=$1
	CNpos2=$2
	((CNposA=CNpos1-10))
	((CNposB=CNpos2+10))
	samtools view A45.A45.bwaAln.st.F12.bam A45:$CNposA-$CNposA | cut -f 1 | sort -u > temp.list1
	samtools view A45.A45.bwaAln.st.F12.bam A45:$CNposB-$CNposB | cut -f 1 | sort -u > temp.list2
	list2_compar.pl temp.list1 temp.list2 temp.list.shared temp.list.u1 temp.list.u2
	wc -l temp.list.shared
	bam_filter_by_readname_file.pl A45.A45.bwaAln.st.F12.bam temp.list.shared 1 A45.A45.bwaAln.st.F12.$CNpos1-$CNpos2.bam
	samtools index A45.A45.bwaAln.st.F12.$CNpos1-$CNpos2.bam
	rm temp.list1 temp.list2 temp.list.shared temp.list.u1 temp.list.u2
	return 0
}


declare -a sparr=("A126" "A38" "A3" "A40" "A45" "A4" "A5" "A68" "A6" "A73" "A7" "A8" "B116" "B117" "B118" "B119" "B120" "B68" "B99")
for indsp in ${sparr[@]};do echo $indsp; fa_error_by_kmer-count.pl -i /home/hpcusers/admin/lufuhao/mtDNA/0.fasta/${indsp}.fa --table /home/hpcusers/admin/lufuhao/mtDNA/k121/${indsp}.k121.dump --mincount 2 --maxcount 255 -k 121 --circular  -a 3 --plot ${indsp}kmc.tab.pdf -o ${indsp}.checkerror; done
```



## Genomic SNP

### mummer method
```
nucmer --maxmatch --delta AABBDD.AABB.AP008982.delta  0.ref/AP008982.fa AABBDD.AABB.fa
delta-filter -q -i 95 AABBDD.AABB.AP008982.delta > AABBDD.AABB.AP008982.qi95.delta 
delta2vcf < AABBDD.AABB.AP008982.qi95.delta > AABBDD.AABB.AP008982.qi95.delta.vcf
```



## BAM
```bash
runBwaAln () {
    RBAindex=$1
    RBAfq1=$2
    RBAfq2=$3
    RBApfx=$4
    RBAopt=$5
    
    bwa aln $RBAopt $RBAindex $RBAfq1 > $RBApfx.R1.sai
    bwa aln $RBAopt $RBAindex $RBAfq2 > $RBApfx.R2.sai
    bwa sampe $RBAindex $RBApfx.R1.sai $RBApfx.R2.sai $RBAfq1 $RBAfq2 | samtools view -b -F4 -h -S - > $RBApfx.unst.bam
    samtools sort -o $RBApfx.st.bam $RBApfx.unst.bam
    samtools index $RBApfx.st.bam
    rm $RBApfx.R1.sai $RBApfx.R2.sai
}
bwa index -p A126bwa A126.fa
runBwaAln A126bwa A126_20210924_FDSW210333331-1r.R1.clean.fq.gz A126_20210924_FDSW210333331-1r.R2.clean.fq.gz A126.bwaAln

```bash
#RunBamClean in.bam
RunBamClean () {
    RBCin=$1
    
    RBCpfx=${RBCin%.bam}
    samtools view -b -h -S -F 4 $RBCin > ${RBCpfx}.F4.bam
    samtools index ${RBCpfx}.F4.bam
    samtools view -b -h -S -F 12 $RBCin > ${RBCpfx}.F12.bam
    samtools index ${RBCpfx}.F12.bam
}
RunBamF12 () {
    RBCin=$1
    RBCpfx=${RBCin%.bam}
    samtools view -b -h -S -F 12 $RBCin > ${RBCpfx}.F12.bam
    samtools index ${RBCpfx}.F12.bam
}

declare -a specarr=("A126" "A38" "A3" "A40" "A45" "A4" "A5" "A68" "A6" "A73" "A7" "A8" "B116" "B117" "B118" "B119" "B120" "B68" "B99")
for idv in ${specarr[@]}; do RunBamF12 A126.bwaAln.st.bam; done
```



## Fragment size

```
insertsize.read.density.rscript -i A68.bedpe -s 'AABBDD Kenong9204' -p A68 --pdf --pointsize 7 --color "red" --xlim 1110 --width 8.5 --height 8.5
insertsize.read.density.rscript -i A73.bedpe -s 'AABBDD AK58' -p A73 --pdf --pointsize 7 --color "red" --xlim 1110 --width 8.5 --height 8.5
insertsize.read.density.rscript -i A126.bedpe -s 'AABBDD Lumai-1' -p A126 --pdf --pointsize 7 --color "red" --xlim 1110 --width 8.5 --height 8.5
```


## Sequencing Depth

```
arr1=('A68' 'A73' 'A126' 'A3' 'A4' 'A7' 'A8' 'A40' 'B116' 'B117' 'B118' 'B119' 'B120' 'A6' 'A45' 'A5' 'B68' 'A38' 'B99')
arr2=('Kenong9204' 'AK58' 'Lumai-1' 'SCAUP' 'DM4' 'Langdon' '81086A' 'Hoh501' 'IX-0944' 'IX-0946' 'IX-0957' 'IX-0969' 'IX-0976' 'Y199' 'AL8/78' 'SQ523' 'Jinan20190529' 'Clae23' 'Kaifeng1')
arr3=('AABBDD' 'AABBDD' 'AABBDD' 'AABB' 'AABB' 'AABB' 'AABB' 'AABB' 'AA' 'AA' 'AA' 'AA' 'AA' 'DD' 'DD' 'DD' 'DD' 'DD' 'DD')
for ((i=0;i<${#arr1[@]}; i++)); do bedgraph.plot.rscript ${arr1[$i]}.bwaAln.st.F12.rmdup.bam.bedgraph "${arr3[$i]} ${arr2[$i]}"; done
```


## Sequence synteny

```
assemdir=/home/lufuhao/P5820T1D2T/temp/mtDNA/0.final_fasta
refdir=/home/lufuhao/P5820T1D2T/temp/mtDNA/0.ref
cat $assemdir/AABBDD/A73.fa $assemdir/AABB/A7.fa $refdir/AP013106.fa $refdir/AP013107.fa $refdir/KJ078648.fa $assemdir/DD/A45.fa $assemdir/DD/B68.fa $assemdir/AA/B116.fa $refdir/MZ202552.fa $refdir/OK120846.fa $refdir/MN127982.fa > query.fa
mum.stat  -q query.fa -r $refdir/AP008982.fa -i 95 -c 300 -l 500 -o interspecies -ps
#gedit interspecies.gp
gnuplot interspecies.gp
```


## Hoh501 double check

```
bamCoverage --bam ../../5.bam/A40.bwaAln.st.F12.rmdup.bam --outFileFormat bedgraph --outFileName A40.bwaAln.st.F12.rmdup.bam.bedgraph --binSize 1 --region "A40:354500:372050"

bedgraph.plot.rscript A40.bwaAln.st.F12.rmdup.bam.bedgraph 'Hoh501:358500-368039' "Hoh501:358500-368039"
```
