# annotation

## GeSeq

## manual curation

> set th word size carefully for short exons

> Find all the pseudogenes >=100bp

```
makeblastdb -in 19.mtDNA.fa -parse_seqids -dbtype nucl -out 19
blastn -query collected.ORF.fa -db 19 -outfmt 6 -word_size 3 -evalue 1E-6 -out collected.ORF.vs.19.blastout
```

## gene alignment

```
genbank_to_fa.py -i $stepdir0/$indseq.gbk -l 60 -o $indseq -f gene -f rRNA -f tRNA -f CDS -f translation > /dev/null
muscle -align $RMin -output $RMout > $RMout.muscle5.log 2>&1
fasta_sort.py -i $indnam.dna.afa -l $indnam.dna.fa.list -o $indnam.dna.fasta -w 60
boxshade -in="$RBSin" -out="$RBSout" -def -ruler $RBSopt -outlen=$RBSwidth -dev=$RBSdev > $RBSout.boxshade.log 2>&1
```

## phylogeny

```
# Align by muscle
# Estimate best models using MEGA > MEDELS > Find Best DNA/protein Models ..., which showed WAG + G + I + F was the best
#MEGA > prototype > save settings > *.mao file
megacc -a infer_ML_amino_acid.mao -d merge.aln.meg -o ML27
```


## Haplotype

> Softwares with UI

```
# DNasP
# Network
```


## OGDraw

> Webservices: https://chlorobox.mpimp-golm.mpg.de/OGDraw.html


## Motif

> remove all * in protein sequence before analysis

```
interproscan.sh -i pep.fasta -seqtype p -appl Pfam -cpu 8 --formats TSV --minsize 6 -iprlookup -o pep.fasta.iprscan
meme -evt 10.0 -nmotifs 10 -minw 6 -maxw 300 -mod anr -oc . -protein pep.fa
### using BioConductor DrawProteins
Rscript motif.rscript atp6.motif
```



## BLAST genome

### BLAST 6 genome
```
export BLASTDB=/home/lufuhao/P5820T1D2T/Databases/blastdb
dbname=('HH' 'AA' 'DD' 'SS' 'AABBt' 'AABBDD')
genename="nad6"
for indb in ${dbname[@]}; do
    if [ ! -s "$genename.vs.$indb.blast6out" ]; then
        blastn -query $genename.dna.w60.fa -evalue 1e-6 -outfmt '6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore qlen slen' -perc_identity 90 -word_size 28 -num_threads 8 -db $indb -out $genename.vs.$indb.blast6out
    fi
    if [ ! -s "$genename.vs.$indb.blast6out.sort" ]; then
        sort -k1,1 -k12,12nr $genename.vs.$indb.blast6out > $genename.vs.$indb.blast6out.sort
    fi
    rm *.blast6out
done
```

